src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/Runners/MonitorRunner.cs

```diff
@@ -240,54 +154,20 @@ public async Task StartAsync(CancellationToken token)
             _outputHelper.WriteLine("User Settings Path: {0}", UserSettingsFilePath);
 
             await _adapter.StartAsync(token);
-
-            using IDisposable _ = token.Register(() => CancelCompletionSources(token));
-
-            // Await ready and exited tasks in case process exits before it is ready.
-            if (_runner.ExitedTask == await Task.WhenAny(_readySource.Task, _runner.ExitedTask))
-            {
-                throw new InvalidOperationException("Process exited before it was ready.");
-            }
-
-            // Await ready task to check if it faulted or cancelled.
-            await _readySource.Task;
-        }
-
-        private void CancelCompletionSources(CancellationToken token)
-        {
-            _defaultAddressSource.TrySetCanceled(token);
-            _metricsAddressSource.TrySetCanceled(token);
-            _readySource.TrySetCanceled(token);
         }
 
-        private void StandardOutputCallback(string line)
+        public virtual async Task WaitForExitAsync(CancellationToken token)
         {
-            ConsoleLogEvent logEvent = JsonSerializer.Deserialize<ConsoleLogEvent>(line);
-
-            switch (logEvent.Category)
+            await Task.Run(() => _runner.ExitedTask.Wait(token)).ConfigureAwait(false);
```

Comment:
I'm not going to do this suggestion, this will cause a leak to occur in the framework whenever the process exits and the cancellation token isn't cancelled. [`DelayPromiseWithCancellation `](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L5625) will fail to call `.Dispose()` on it's `CancellationTokenRegistration`

Reply:
Oh, because of the infinite timeout. That's really easy to fix though... just use some absurdly large timeout value or a relatively smallish one and then do:

```cs
Task timeoutTask = Task.Delay(nonInfiniteTimeout, token);
Task completedTask = await Task.WhenAny(_runner.ExitedTask, timeoutTask).ConfigureAwait(false);
token.ThrowIfCancellationRequested();
if (completedTask == timeoutTask) throw new TimeoutException();
```

but the other suggestion still stands as being simpler:

```cs
await _runner.WaitForExitAsync(token).ConfigureAwait(false);
```

Reply:
I figured out a way to make this work and work around the problem with the leak from an infinite waiting task. I also made it an extension method since it's pretty verbose.

Reply:
On second though, you're probably right, we should throw if the cancellation token gets fired.